Automattic/wp-calypso
# Combined calypso apps

## Proposed Changes
Build and process all calypso apps in parallel, in a single TeamCity build, rather than independent builds. This significantly improves resource usage in TeamCity. Importantly, merging this PR **will not change the process on trunk**. After it's merged, we can integrate the changes with `install-plugin.sh`, and then we can remove the old builds. So the old builds continue working as they did, with some minor tweaks to handle the required changes. 

Corresponding diff for `install-plugin.sh` changes: D125086-code

## Design:
Each Calypso App has a normal build script. However, in TeamCity, more things might need to happen to produce the final deployable artifact. For some apps, assets are copied into a different directory. Others use `sed` to insert a version constant in a PHP file.

To help with this, every app now has a `teamcity:build-app` script. This allows each app to execute what it needs to do for TeamCity. For example:
- It could run `yarn build`
- It could run `yarn build && yarn run check-size` (to perform an extra test in CI)
- It could run `./bin/ci-build-steps.sh` (to do a bunch of stuff in bash)

The bash script is only really used for the PHP plugins. In ETK, for example, this is used to update the version constant based on the TeamCity build number. For happy blocks and o2-blocks, this is used to copy things to a different build directory (which is admittedly an anti-pattern, since that _should_ be done by webpack, but out of scope for this PR).

All of the teamcity:build-app scripts are executed via yarn workspaces, and TeamCity automatically builds all projects under the `apps` directory. (Which makes it easier to integrate new apps into the process.)

The step which runs `teamcity:build-app` generates all of the final artifacts in parallel.

Then, the "process app artifacts script" starts. This generates a list of apps (based on the directories in /apps), and does the a few steps for each app in parallel:
1. Download the previous release artifact for that app.
2. Execute any `bin/normalize-artifact.sh` script found for that app. (To remove anything that breaks idempotence, like new version numbers or hashes. This is only needed for a couple of apps currently, but any app can define the script to have it executed.)
3. Compare the previous release artifact to the one generated in the build step above.

Now we have a list of apps that "changed". For each of these, we run existing logic (implemented in #73557), which can send Slack notifications and update GitHub comments.

I also updated the "calypso app builder" package to write README and build_meta files. Previously (on trunk), there were two different approaches. For some apps which have only one webpack build, they would use a webpack plugin to write build meta. For others with multiple builds (like ETK), a build meta txt file was created by TeamCity. Additionally, the README was copied into the release artifact by TeamCity to improve documentation in the wpcom repo.

The new approach writes a build_meta JSON file (in the same format as the webpack plugin), but in the "calypso apps builder" script instead. I also moved the "copy readme" step into that script. It already coordinates all the webpack builds for an individual app, so it's a good "source of truth" for an app's build logic. This means all the new release artifacts have working build_meta & README files, all using consistent logic.

## Testing Instructions

1. Verify the build works as expected here: https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WPComPlugins_Build_Plugins?mode=builds
2. Verify that the existing build results are not changing. (For example, the build inside the combined calypso apps should be nearly identical to the independent build, and the independent build in this branch should match trunk for the same base commit.)
